src/Decimal.php

```diff
@@ -649,53 +711,95 @@ protected function copy(?int $integerPart = null, ?string $decimalPart = null, ?
      * @param string $value
      * @param int|null $scale
      *
-     * @throws \InvalidArgumentException
-     *
      * @return void
      */
     protected function setValue(string $value, ?int $scale): void
     {
-        preg_match('#(.+)e(.+)#i', $value, $matches);
-        if (!$matches) {
-            $separatorPos = strpos($value, '.');
-            if ($separatorPos !== false) {
-                $before = (int)substr($value, 0, $separatorPos);
-                $after = substr($value, $separatorPos + 1);
-            } else {
-                $before = (int)$value;
-                $after = '';
-            }
+        if (preg_match('#(.+)e(.+)#i', $value) === 1) {
+            $this->fromScientific($value, $scale);
 
-            $this->negative = $before < 0 || ($before === 0 && $after !== '' && strpos($value, '-') === 0);
-            $this->integralPart = abs($before);
-            $this->fractionalPart = $after;
+            return;
+        }
+
+        if (strpos($value, '.') !== false) {
+            $this->fromFloat($value);
 
             return;
         }
 
+        $this->fromInt($value);
+    }
+
+    /**
+     * @param string $value
+     *
+     * @throws \InvalidArgumentException
+     *
+     * @return void
+     */
+    protected function fromInt(string $value): void
+    {
+        preg_match('/^(-)?([^.]+)$/', $value, $matches);
+        if ($matches === []) {
+            throw new InvalidArgumentException('Invalid integer number');
+        }
+
+        $this->negative = $matches[1] === '-';
+        $this->integralPart = $matches[2];
+        $this->fractionalPart = '';
+    }
+
+    /**
+     * @param string $value
+     *
+     * @throws \InvalidArgumentException
+     *
+     * @return void
+     */
+    protected function fromFloat(string $value): void
+    {
+        preg_match('/^(-)?(.*)\.(.*)$/', $value, $matches);
+        if ($matches === []) {
+            throw new InvalidArgumentException('Invalid float number');
+        }
+
+        $this->negative = $matches[1] === '-';
+        $this->integralPart = $matches[2];
+        $this->fractionalPart = $matches[3];
+    }
+
+    /**
+     * @param string $value
+     * @param int|null $scale
+     *
+     * @throws \InvalidArgumentException
+     *
+     * @return void
+     */
+    protected function fromScientific(string $value, ?int $scale): void
+    {
         $pattern = '/^(-?)(\d+(?:' . static::RADIX_MARK . '\d*)?|' .
             '[' . static::RADIX_MARK . ']' . '\d+)' . static::EXP_MARK . '(-?\d*)?$/i';
         preg_match($pattern, $value, $matches);
         if (!$matches) {
-            throw new InvalidArgumentException('Invalid value/notation: ' . $value);
+            throw new InvalidArgumentException('Invalid scientific value/notation: ' . $value);
         }
 
-        $negativeChar = $matches[1];
-        $value = $matches[2];
-        $floatValue = (float)$value;
```

Comment:
`float` can't really be trusted, for example `(float)PHP_INT_MAX` is `double(9.2233720368548E+18)` while `PHP_INT_MAX` is `9223372036854775807`, not even the same number.